### ISSUE TITLE
Help wanted: Improving performance when repeatedly creating a one-entry bibliography

### ISSUE THREAD

**Issue by halirutan** [michel-kraemer/citeproc-java#40]
In JabRef, we use 

```
CSL.makeAdhocBibliography(style, outputFormat.getFormat(), cslItemData);
```

to create the HTML-preview for the currently selected item in the BibTeX database. Every time the user scrolls through the main list, we update the preview and call this code again. I have [carefully profiled what happens](https://github.com/JabRef/jabref/issues/2533#issuecomment-350601001), but everything points to this call being the source of our memory and performance problems. I know that `makeAdhocBibliography` has a clear warning sign in its comments that it will be expensive.

Can you recommend the best possible way to handle our situation so that it is as resource-friendly as possible? We have the following constants and variables:

- the style is set in the preferences and stays constant most of the time
- the output format is fixed as well
- we provide the `cslItemData` every time the preview needs to be updated

I'm asking this question because at the moment I see no sensible way with one of the built-in `ItemDataProviders` to achieve this. Therefore, one of my first questions would be: Is it OK to make an `ItemDataProvider` that changes its content? I would then initialize a `CSL` instance with an empty provider and change its underlying `ItemData` list every time we need to render an item.

**Comment by michel-kraemer** [comment]
Your observations are correct. `makeAdhocBibliography` is good if you only need to create a bibliography once or every now and then, but if you need to create multiple bibliographies you should try to reuse the `CSL` object. The reason for this is that the `CSL` object holds a `ScriptRunner` which is very expensive to create. AFAIR the library contains code that automatically tries to reuse `ScriptRunner` instances by keeping them in a `ThreadLocal` but I guess this depends on your setup and on the actual JavaScript engine used.

Regarding your problem with the built-in `ItemDataProvider`: Just implement your own one that provides citation items dynamically (without a static list like the `ListItemDataProvider` does). This should work. I think you only need to implement two methods: `retrieveItem` which is only called at the moment the bibliography is created, and `getIds` which should return an array of all available citation IDs. I don't remember how often `getIds` is called and whether you could return a dynamic result here, but I guess you should just try it out. An alternative would be to only re-create the `CSL` object if the user's list of citation items has changed.

In addition, @koppor is right when he says you should try V8. By default citeproc-java uses Nashorn and this engine is far from being fast, especially at the beginning when it's not warm yet and the JIT compiler has not kicked in (another reason why you should try to keep the `CSL` instance in memory. The more often you execute the same JavaScript code the better the results of the JIT compiler).

I hope this helps. Let me know if you need more assistance. I can try to create some sample code if you like. Just give me more details, so I can better reproduce your scenario.

**Comment by halirutan** [comment]
Thank you, Michel, for this long answer. I wrote an adapter-class for your CSL engine that holds an instance and synchronizes access to it. Using `ThreadLocal` was difficult since we run this as `SwingWorker` and cannot really control which thread it gets assigned to.
I tried it locally and even with fast scrolling through the entries, where I'm sure that several threads try to create a preview, works like a charm. It is soooo much faster.

Thanks again for your time.
